pkg/cloudprovider/aws/initialization.go

```diff
@@ -0,0 +1,242 @@
+		log.PanicIfError(aerr, "Failed to retrieve default launch template")
+	}
+	if len(describeLaunchTemplateOutput.LaunchTemplates) > 0 {
+		zap.S().Infof("Successfully detected existing launch template, %s/%s",
+			*describeLaunchTemplateOutput.LaunchTemplates[0].LaunchTemplateId,
+			*describeLaunchTemplateOutput.LaunchTemplates[0].LaunchTemplateName)
+		return describeLaunchTemplateOutput.LaunchTemplates[0]
+	}
+
+	// 2. Create Launch Template
+	createLaunchTemplateOutput, err := ec2API.CreateLaunchTemplate(&ec2.CreateLaunchTemplateInput{
+		LaunchTemplateName: aws.String(name),
+		LaunchTemplateData: &ec2.RequestLaunchTemplateData{
+			IamInstanceProfile: &ec2.LaunchTemplateIamInstanceProfileSpecificationRequest{
+				Name: aws.String(instanceProfileName),
+			},
+			TagSpecifications: []*ec2.LaunchTemplateTagSpecificationRequest{{
+				ResourceType: aws.String(ec2.ResourceTypeInstance),
+				Tags: []*ec2.Tag{{
+					Key:   aws.String(fmt.Sprintf("kubernetes.io/cluster/%s", *cluster.Name)),
+					Value: aws.String("owned"),
+				}},
+			}},
+			SecurityGroupIds: append(cluster.ResourcesVpcConfig.SecurityGroupIds, cluster.ResourcesVpcConfig.ClusterSecurityGroupId),
+			UserData: aws.String(base64.StdEncoding.EncodeToString([]byte(fmt.Sprintf(`
+				#!/bin/bash
+				yum install -y https://s3.amazonaws.com/ec2-downloads-windows/SSMAgent/latest/linux_amd64/amazon-ssm-agent.rpm
+				/etc/eks/bootstrap.sh %s \
+					--kubelet-extra-args '--node-labels=karpenter.sh/provisioned=true' \
+					--b64-cluster-ca %s \
+					--apiserver-endpoint %s`,
+				*cluster.Name,
+				*cluster.CertificateAuthority.Data,
+				*cluster.Endpoint,
+			)))),
+			ImageId: aws.String("ami-0532808ed453f9ca3"),
+		},
+	})
+	log.PanicIfError(err, "Failed to create default launch template")
+	zap.S().Infof("Successfully created default launch template, %s/%s",
+		*createLaunchTemplateOutput.LaunchTemplate.LaunchTemplateId,
+		*createLaunchTemplateOutput.LaunchTemplate.LaunchTemplateName)
+	return createLaunchTemplateOutput.LaunchTemplate
+}
+
+func zonalSubnetsOrDie(ec2API ec2iface.EC2API, cluster *eks.Cluster) map[string]*ec2.Subnet {
+	describeSubnetOutput, err := ec2API.DescribeSubnets(&ec2.DescribeSubnetsInput{
+		SubnetIds: cluster.ResourcesVpcConfig.SubnetIds,
+	})
+	log.PanicIfError(err, "Failed to describe subnets %v", cluster.ResourcesVpcConfig.SubnetIds)
+	zonalSubnetMap := map[string]*ec2.Subnet{}
+	// TODO Filter public subnets
+	for _, subnet := range describeSubnetOutput.Subnets {
+		zonalSubnetMap[*subnet.AvailabilityZone] = subnet
+	}
+
+	return zonalSubnetMap
+}
+
+func ensureAWSAuthOrDie(config *rest.Config, role *iam.Role) {
+	kubeClient, err := client.New(config, client.Options{})
+	log.PanicIfError(err, "Failed to instantiate ephemeral kubeClient")
+
+	awsAuth := &v1.ConfigMap{}
+	nn := types.NamespacedName{Name: "aws-auth", Namespace: "kube-system"}
+	log.PanicIfError(kubeClient.Get(context.TODO(), nn, awsAuth), "Failed to retrieve configmap aws-auth")
```

Comment:
IMO, its a little hard to read, can we keep function call separate and pass in err to the log.

Reply:
Good point